presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -101,89 +104,113 @@ public static PlanNodeStatsEstimate filterStatsForUnknownExpression(PlanNodeStat
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitExpression(Expression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitExpression(Expression node, Void context)
         {
-            return filterForUnknownExpression();
-        }
-
-        private PlanNodeStatsEstimate filterForUnknownExpression()
-        {
-            return filterStatsForUnknownExpression(input);
+            return Optional.empty();
         }
 
-        private PlanNodeStatsEstimate filterForFalseExpression()
+        private Optional<PlanNodeStatsEstimate> filterForFalseExpression()
         {
             PlanNodeStatsEstimate.Builder falseStatsBuilder = PlanNodeStatsEstimate.builder();
             input.getSymbolsWithKnownStatistics().forEach(symbol -> falseStatsBuilder.addSymbolStatistics(symbol, ZERO_STATS));
-            return falseStatsBuilder
+            return Optional.of(falseStatsBuilder
                     .setOutputRowCount(0.0)
-                    .build();
+                    .build());
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitNotExpression(NotExpression node, Void context)
         {
-            return differenceInStats(input, process(node.getValue()));
+            return process(node.getValue()).map(childStats -> differenceInStats(input, childStats));
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
         {
-            PlanNodeStatsEstimate leftStats = process(node.getLeft());
-            PlanNodeStatsEstimate andStats = new FilterExpressionStatsCalculatingVisitor(leftStats, session, types).process(node.getRight());
-
             switch (node.getType()) {
                 case AND:
-                    return andStats;
+                    return visitLogicalBinaryAnd(node.getLeft(), node.getRight());
                 case OR:
-                    PlanNodeStatsEstimate rightStats = process(node.getRight());
-                    PlanNodeStatsEstimate sumStats = addStatsAndSumDistinctValues(leftStats, rightStats);
-                    return differenceInNonRangeStats(sumStats, andStats);
+                    return visitLogicalBinaryOr(node.getLeft(), node.getRight());
                 default:
                     throw new IllegalStateException("Unimplemented logical binary operator expression " + node.getType());
             }
         }
 
+        private Optional<PlanNodeStatsEstimate> visitLogicalBinaryAnd(Expression left, Expression right)
+        {
+            Optional<PlanNodeStatsEstimate> leftStats = process(left);
+            if (leftStats.isPresent()) {
```

Comment:
Couldn't we just check if returned stats are `UNKNOWN_STATS` (if we are already replacing `UNKNOWN_STATS` by `Optional.empty()` as in https://github.com/prestodb/presto/pull/10384/commits/92d70e10865dee1a7387ccce3119da2515dc565a#diff-2c8e7079274d1e50a57e44770854b2c5R297)